Add infrastructure to rewrite query builders

s1monw opened the pull request:
QueryBuilders today do all their heavy lifting in toQuery() which
can be too late for several operations. For instance if we want to fetch geo shapes
on the coordinating node we need to do all this before we create the actual lucene query
which happens on the shard itself. Also optimizations for request caching need to be done
to the query builder rather than the query which then in-turn needs to be serialized again.
This commit adds the basic infrastructure for query rewriting and moves the heavy lifting into
the rewrite method for the following queries:

 * `WrapperQueryBuilder`
 * `GeoShapeQueryBuilder`
 * `TermsQueryBuilder`
 * `TemplateQueryBuilder`

Other queries like `MoreLikeThisQueryBuilder` still need to be fixed / converted. The nice
sideeffect of this is that queries like template queries will now also match the request cache
if their non-template equivalent has been cached befoore. In the future this will allow to
add optimizataion like rewriting time-based queries into primitives like `match_all_docs` or `match_no_docs`
based on the currents shards bounds. This is especially appealing for indices that are read-only ie. never change.

s1monw commented:
@colings86 @MaineC @cbuescher @javanna this would have been a damn disaster without all the tests you folks added for queries. Thanks so much for this, the reward for all the pain is huge since you enabled others to make massive progress in no time! Worth all the pain I hope!

colings86 commented on core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java:
```diff
@@ -227,22 +226,13 @@ public String getWriteableName() {
-        } else {
-            terms = values;
+        if (termsLookup != null) {
+            throw new UnsupportedOperationException("query must be rewritten first");
```
nit: should this be IllegalStateException rather than UOE since the method is valid to call but in this case the class is not in the right state?

colings86 replied:
Actually ignore this, I didn't see that this query gets rewritten into a different query when there is a termsLookup

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -346,4 +347,40 @@ protected void doWriteTo(StreamOutput out) throws IOException {
+            if (result != builder) {
+                changed = true;
+            }
+            conusmer.accept(result);
```
s/conusmer/consumer

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryBuilder.java:
```diff
@@ -72,4 +72,28 @@
     String getName();
+
+    /**
+     * Rewrites this query builder into it's primitive form. By default this method return theb builder itself. If the builder
```
s/it's primitive form/its primitive form  and  s/theb/the

colings86 commented on core/src/main/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryBuilder.java:
```diff
@@ -394,4 +395,17 @@ public FilterFunctionBuilder readFrom(StreamInput in) throws IOException {
+    @Override
+    public QueryBuilder<?> rewrite(QueryRewriteContext queryShardContext) throws IOException {
+        QueryBuilder<?> queryBuilder = this.query.rewrite(queryShardContext);
+        if (queryBuilder != query) {
```
Can the functions not have a filter as well? Should we rewrite the functions too?

javanna replied:
good point! I think we need to iterate over the filterFunctionBuilders and rewrite their corresponding filters

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryBuilder.java:
```diff
@@ -72,4 +72,28 @@
+
+    /**
+     * Rewrites the given query into it's primitive form. Queries that for instance fetch resources from remote hosts or
+     * can simplify / optimize itself should do their heavy lifting duringt {@link #rewrite(QueryRewriteContext)}. This method
```
s/it's primitive form/its primitive form    s/itself/themselves     s/duringt/during

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java:
```diff
@@ -0,0 +1,63 @@
+import org.elasticsearch.script.ScriptService;
+
+/**
+ */
```
javadocs please? :)

colings86 commented:
@s1monw I left a couple of comments but I really like this change :)

javanna commented on core/src/main/java/org/elasticsearch/index/query/ScriptQueryBuilder.java:
```diff
@@ -32,6 +33,7 @@
 import org.elasticsearch.script.Script;
 import org.elasticsearch.script.Script.ScriptField;
 import org.elasticsearch.script.ScriptContext;
+import org.elasticsearch.script.ScriptEngineService;
```
unused imports?

javanna commented on core/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -33,6 +33,7 @@
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.index.IndexNotFoundException;
 import org.elasticsearch.indices.IndexClosedException;
+import org.elasticsearch.search.SearchService;
```
unused import?

javanna commented on core/src/main/java/org/elasticsearch/index/query/EmptyQueryBuilder.java:
```diff
@@ -23,10 +23,12 @@
 import org.elasticsearch.action.support.ToXContentToBytes;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ToXContent;
```
this is unused, as well as XContentType and ToXContentToBytes

javanna commented on core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java:
```diff
@@ -105,14 +105,7 @@ public String getWriteableName() {
-            context.combineNamedQueries(contextCopy);
-            return result.toQuery(context);
-        }
+       throw new UnsupportedOperationException("this query must be rewritten first");
```
fix the indentation?

javanna commented on modules/lang-mustache/src/test/java/org/elasticsearch/messy/tests/TemplateQueryParserTests.java:
```diff
@@ -213,8 +214,23 @@ public void testParserCanExtractTemplateNames() throws Exception {
+        templateSourceParser.nextToken();
+        TemplateQueryParser parser = injector.getInstance(TemplateQueryParser.class);
+        try {
+            parser.fromXContent(context.parseContext()).toQuery(context);
```
missing `fail()` here ?

javanna commented on modules/lang-mustache/src/test/java/org/elasticsearch/messy/tests/TemplateQueryParserTests.java:
```diff
@@ -213,8 +214,23 @@ public void testParserCanExtractTemplateNames() throws Exception {
         assertTrue("Parsing template query failed.", query instanceof MatchAllDocsQuery);
     }
+
+    public void testMustRewrite() throws Exception {
```
how about adding this type of test to all the queries where we throw exception as part of toQuery if rewrite wasn't called? or should we in general enforce that rewrite must be always called before calling toQuery?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:
```diff
@@ -73,15 +68,13 @@
  * Context object used to create lucene queries on the shard level.
  */
-public class QueryShardContext {
+public class QueryShardContext extends QueryRewriteContext {
```
is it necessary that `QueryShardContext` extends `QueryRewriteContext`. I see a lot of parse stuff around but at this point this context shouldn't need to parse anything anymore? Sure I am missing something :)

s1monw replied:
the obvious thing. it's a true subclass of QueryRewriteContext it shares all it's members it's just more powerful. If you have a `QueryShardContext` there is no need to create another instance you can just pass it on?

javanna replied:
Sure I see where this is handy, but from a design perspective, does the `QueryShardContext` ever need to parse anymore? or does it need the ScriptService? I think it may be convenient to isolate those to make sure that `toQuery` doesn't do any parsing/scripting at this point?

s1monw replied:
yes it does, ScriptQuery, MLT etc. as I stated in the description

javanna replied:
boooo ok I was hoping we had moved all that to rewrite.

s1monw replied:
ScriptQuery might not be possible or only limitied but MLT just needs work

javanna commented on core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java:
```diff
@@ -552,6 +552,12 @@ public void testToQuery() throws IOException {
 
+    private QueryBuilder<?> rewriteQuery(QB queryBuilder, QueryRewriteContext rewriteContext) throws IOException {
+        QueryBuilder<?> rewritten = QueryBuilder.rewriteQuery(queryBuilder, rewriteContext);
+        assertSerialization(rewritten);
```
question: what does it buy us to call `assertSerialization` here if every query has specific serialization tests that already call it? Afraid of forgotten queries, or what else?

s1monw replied:
will add a comment, this is basically testing the serialization of a rewritten query to ensure that the rewritten version is also serializable etc.

javanna replied:
I see what it is doing, I was wondering why we are doing it given that every single query (that we might rewrite to) is already tested. More coverage? Potentially untested queries?

s1monw replied:
if you rewrite into a state that is not serialized you are fucked and will never detect it

javanna commented on core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java:
```diff
@@ -552,6 +552,12 @@ public void testToQuery() throws IOException {
         }
     }
 
```
at the moment we test rewrite implicitly as we call it all the time right before calling toQuery. I think rewrite may deserve to be tested separately though? Just to make sure that it does what it's supposed to do. E.g. function score that rewrites its function filters too etc. otherwise we only test the input and the output but not the intermediate format. What do you think?

s1monw replied:
I can add a test for the ones that rewrite

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -106,7 +106,7 @@ public boolean isDeprecatedSetting(String setting) {
         if (token == XContentParser.Token.END_OBJECT) {
             // empty query
-            return EmptyQueryBuilder.PROTOTYPE;
+            return new EmptyQueryBuilder();
```
for my education, was there any problem with the previous approach?

s1monw replied:
so now that we rewrite we potentially return an EmptyQueryBuilder form a rewrite method. for instance with `new WrapperQueryBuilder("{}")` this is problematic if the `WrapperQueryBuilder` had a query name which is inherited but ignored by the old version to make all assumptions hold we have to upgrade this query builder to a proper one., makes sense?

javanna replied:
I see, cause the empty query builder needs to be modified. Cool thanks.

javanna commented on core/src/main/java/org/elasticsearch/index/query/EmptyQueryBuilder.java:
```diff
@@ -37,75 +39,49 @@
 
     public static final String NAME = "empty_query";
 
     /** the one and only empty query builder */
```
comment is outdated if we go for creating a new instance all the time

javanna commented on core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java:
```diff
@@ -134,4 +127,21 @@ protected int doHashCode() {
+            parseContext.reset(qSourceParser);
+
+            final QueryBuilder<?> queryBuilder = parseContext.parseInnerQueryBuilder();
+            if (queryBuilder.boost() == DEFAULT_BOOST) {
```
I think this is not correct, as we will ignore one of the boosts. previously the two boosts 9if they both had a value != 1.0f) caused the queries to be wrapped twice with BoostQuery, hence the two boosts ended up being multiplied by lucene. I think we may want to multiply ourselves here and in WrapperQueryBuilder.

javanna commented on core/src/main/java/org/elasticsearch/index/query/WrapperQueryBuilder.java:
```diff
@@ -134,4 +127,21 @@ protected int doHashCode() {
+            }
+            if (queryName() != null) { // we inherit the name
+                queryBuilder.queryName(queryName());
+            }
```
do we want to move the boost logic and query name to AbstractQueryBuilder somehow? I am always afraid of forgetting these tiny details.

javanna commented on core/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java:
```diff
@@ -237,12 +236,11 @@ public ShapeRelation relation() {
 
     @Override
     protected Query doToQuery(QueryShardContext context) throws IOException {
```
if we don't do any parsing as part of toQuery, I think the throws clause can go away from its signature everywhere, same for doToQuery

javanna replied:
seems like we cannot do this just because of HasChildQueryBuilder and its parsing of inner_hits... bla

javanna commented:
I left a bunch of comments, mainly minors, but this looks very good, I like the change. One question: we are not moving yet to calling rewrite on the coordinating node, thus fetching shapes and terms is still on the data nodes, but is that the plan on the long run? Cause when we discussed that in the past, we weren't sure if it was a good idea to move the fetch to the coord node and then serialize fetched terms and shapes from coord node to the data nodes. Curious if anything has changed there.

javanna commented on core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java:
```diff
@@ -625,11 +631,13 @@ public void testSerialization() throws IOException {
      */
     @SuppressWarnings("unchecked")
-    protected QB assertSerialization(QB testQuery) throws IOException {
+    protected <QB extends QueryBuilder> QB assertSerialization(QB testQuery) throws IOException {
```
why is `<QB extends QueryBuilder>` needed? it hides `QB` that's already defined in this class.

s1monw replied:
since the QB type that is passed in can be different to the QB type of the class

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:
```diff
@@ -426,12 +395,16 @@ public ParsedQuery toQuery(QueryBuilder<?> queryBuilder) {
+            return Queries.newMatchNoDocsQuery();
         }
         return query;
     }
```
the combineNamedQueries method remains unused now, can you remove it?

s1monw replied:
done

s1monw commented:
@colings86 @javanna I pushed new commits addressing all comments.

> I left a bunch of comments, mainly minors, but this looks very good, I like the change. One question: we are not moving yet to calling rewrite on the coordinating node, thus fetching shapes and terms is still on the data nodes, but is that the plan on the long run? Cause when we discussed that in the past, we weren't sure if it was a good idea to move the fetch to the coord node and then serialize fetched terms and shapes from coord node to the data nodes. Curious if anything has changed there.

I don't know it was just an example of how this can be used. I am more focusing on rewriting before stuff is added to the percolator etc. in the future.

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -262,6 +263,13 @@ public String getWriteableName() {
+            // named queries TemplateQueryBuilder and WrapperQueryBuilder.
+            // it's equivalent but will anyways happen later down the road in the BQ#rewrite method
+            return mustClauses.get(0).toQuery(context);
+        }
```
should we instead move this special case to TemplateQueryBuilderTests and WrapperQueryBuilderTests doAssertLuceneQuery so that they assert on the right thing?

javanna commented on core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java:
```diff
@@ -948,4 +976,16 @@ private static String msg(String left, String right) {
+     * This test ensures that queries that need to be rewritten have dedicated tests.
+     * These queries must override this method accordingly.
+     */
+    public void testMustRewrite() throws IOException {
```
This one makes FunctionScoreQueryBuilderTests with this seed: 5FB367958E3049EF:EB5C14A8F6674F6F . that's because although you set allow unmapped fields to the context, function score needs to have the field mapped (see DecayFunctionBuilder#187).

javanna commented:
left 2 minors, LGTM otherwise!

javanna commented on core/src/test/java/org/elasticsearch/index/query/TermsQueryBuilderTests.java:
```diff
@@ -272,5 +272,17 @@ public void testFromJson() throws IOException {
     }
+
+    @Override
+    public void testMustRewrite() throws IOException {
```
this one fails with seed `388D83607E509370`

s1monw replied:
fixed this one too... termslookup doesn't pass on null value

s1monw commented:
@javanna pushed new commit

javanna commented:
LGTM
